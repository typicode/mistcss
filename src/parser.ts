import {
  compile,
  DECLARATION,
  Element,
  Middleware,
  middleware,
  COMMENT,
  RULESET,
  // @ts-ignore
  SCOPE,
  serialize,
} from 'stylis'

export interface Data {
  comment: string
  className: string // foo
  tag: string // div
  attributes: Record<string, Set<string>> // data-foo: ['bar', 'baz']
  booleanAttributes: Set<string> // data-foo, data-bar
  properties: Set<string> // --foo, --bar
}

function getComment(element: Element): string {
  // @ts-ignore
  const prev = element.siblings[element.siblings.indexOf(element) - 1] as Element | undefined
  if (prev?.type === COMMENT) {
    return (prev.children as string).trim()
  }
  return ''
}

// (div.foo) -> { tag: 'div', className: 'foo' }
function parseScopeSelector(str: string): { tag: string; className: string } {
  const [tag = '', className = ''] = str.slice(1, -1).split('.')
  return { tag, className }
}

// :scope[data-foo="bar"] -> true
// :scope[data-foo] -> true
export function isAttribute(str: string): boolean {
  return str.startsWith(':scope[data-')
}

// :scope[data-foo="bar"] -> { attribute: 'data-foo',  value: 'bar' }
// :scope[data-foo] -> { attribute: 'data-foo',  value: '' }
function parseAttribute(str: string): { attribute: string; value?: string } {
  const [attribute = '', value] = str.slice(7, -1).split('=')
  return { attribute, value: value?.slice(1, -1) }
}

function update(data: Data): Middleware {
  return function(element, _index, _children, callback) {
    switch (element.type) {
      case DECLARATION:
        // Custom properties
        if ((element.props as string).startsWith('--')) {
          data.properties.add(element.props as string)
        }
        break

      case RULESET:
        ; (element.props as string[])
          .filter(isAttribute)
          .map(parseAttribute)
          .forEach(({ attribute, value }) => {
            if (value === undefined) {
              data.booleanAttributes.add(attribute)
            } else {
              const set = (data.attributes[attribute] ??= new Set())
              set.add(value)
            }
          })
        break
    }

    Array.isArray(element.children) && serialize(element.children, callback)
  }
}

export function parse(css: string): Data[] {
  const parsed: Data[] = []

  serialize(
    compile(css),
    middleware([
      (element) => {
        // Seach for @scope
        switch (element.type) {
          case SCOPE: {
            const prop = element.props[0] // (.foo)
            if (prop === undefined) return // Not supported
            const { tag, className } = parseScopeSelector(prop)

            const data: Data = {
              comment: getComment(element),
              tag,
              className,
              attributes: {},
              booleanAttributes: new Set<string>(),
              properties: new Set<string>(),
            }
            parsed.push(data)

            // Extract data-attributes and custom properties
            serialize(
              // Loop through @scope children
              // Note: nested rules are children of @scope, not :scope
              element.children as Element[],
              middleware([update(data)]),
            )
          }
        }
      },
    ]),
  )

  return parsed
}
