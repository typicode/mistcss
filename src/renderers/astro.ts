import { attributeToCamelCase, propertyToCamelCase } from './_case.js'
import { Data } from '../parser.js'
import { renderTag, renderPropsInterface } from './_common.js'

function renderProps(data: Data): string {
  return [
    'const {',
    [
      ...Object.keys(data.attributes).map(attributeToCamelCase),
      ...Array.from(data.booleanAttributes).map(attributeToCamelCase),
      ...Array.from(data.properties).map(propertyToCamelCase),
      '...props',
    ].join(', '),
    '} = Astro.props',
  ].join(' ')
}

// https://docs.astro.build/en/guides/typescript/
export function render(filename: string, data: Data): string {
  return `---
// Generated by MistCSS, do not modify
import type { HTMLAttributes } from 'astro/types'
import './${filename}.mist.css'

${renderPropsInterface(data, `HTMLAttributes<'${data.tag}'>`)}

${renderProps(data)}
---

${renderTag(data, '<slot />', 'class')}
`
}
