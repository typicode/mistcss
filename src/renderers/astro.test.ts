import assert from 'node:assert/strict'
import test from 'node:test'

import { Data } from '../parser.js'
import { render } from './astro.js'

void test('render', () => {
  const data: Data = {
    tag: 'div',
    className: 'foo',
    attributes: {
      'data-attr': new Set(['a', 'b']),
      'data-attr-foo-bar': new Set(['foo-bar']),
    },
    booleanAttributes: new Set(['data-is-foo']),
    properties: new Set(['--prop-foo', '--prop-bar']),
  }

  const actual = render('component', data)
  assert.equal(
    actual,
    `---
// Generated by MistCSS, do not modify
import './component.mist.css'

interface Props {
  attr?: 'a' | 'b'
  attrFooBar?: 'foo-bar'
  isFoo?: boolean
  propFoo?: string
  propBar?: string
}

const { attr, attrFooBar, isFoo, propFoo, propBar, ...rest } = Astro.props
---

<div
  {...rest}
  data-attr={attr}
  data-attr-foo-bar={attrFooBar}
  data-is-foo={isFoo}
  style={{ '--prop-foo': propFoo, '--prop-bar': propBar }}
  class="foo"
>
  <slot />
</div>
`,
  )
})


void test('render', () => {
  const data: Data = {
    tag: 'hr',
    className: 'foo',
    attributes: {},
    booleanAttributes: new Set(),
    properties: new Set()
  }

  const actual = render('component', data)
  assert.equal(
    actual,
    `---
// Generated by MistCSS, do not modify
import './component.mist.css'

interface Props {

}

const { ...rest } = Astro.props
---

<hr
  {...rest}

  style={{  }}
  class="foo"
/>
`,
  )
})

