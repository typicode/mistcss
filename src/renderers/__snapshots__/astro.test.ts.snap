// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render > renders Astro component (full) 1`] = `
"---
// Generated by MistCSS, do not modify
import type { HTMLAttributes } from 'astro/types'
import './component.mist.css'

type Props = { attr?: 'a' | 'b', attrFooBar?: 'foo-bar', isFoo?: boolean, propFoo?: string, propBar?: string } & HTMLAttributes<'div'>

const { attr, attrFooBar, isFoo, propFoo, propBar, ...props } = Astro.props
---

<div {...props} data-attr={attr} data-attr-foo-bar={attrFooBar} data-is-foo={isFoo} style={{ '--prop-foo': propFoo, '--prop-bar': propBar }} class="foo" ><slot /></div>
"
`;

exports[`render > renders Astro component (minimal) 1`] = `
"---
// Generated by MistCSS, do not modify
import type { HTMLAttributes } from 'astro/types'
import './component.mist.css'

type Props = {  } & HTMLAttributes<'div'>

const { ...props } = Astro.props
---

<div {...props} class="foo" ><slot /></div>
"
`;

exports[`render > renders Astro component (void element) 1`] = `
"---
// Generated by MistCSS, do not modify
import type { HTMLAttributes } from 'astro/types'
import './component.mist.css'

type Props = {  } & HTMLAttributes<'hr'>

const { ...props } = Astro.props
---

<hr {...props} class="foo" />
"
`;
